src/cargo/core/features.rs

```diff
@@ -627,7 +653,32 @@ impl CliUnstable {
         }
         let mut warnings = Vec::new();
         for flag in flags {
-            self.add(flag, &mut warnings)?;
+            if flag.starts_with("allow-features=") {
+                self.add(flag, &mut warnings)?;
+            }
+        }
+        if let Some(allowed) = self.allow_features.take() {
```

Comment:
Instead of taking the allowed set and reinserting it, can this be changed so that the check is in the `add` method?  That might help simplify this.  That is, remove this condition and just keep the `else` clause, and add something like this to `add`:

```rust
if let Some(allowed) = &self.allow_features {
    if k != "allow-features" && !allowed.contains(k) {
        bail!("the feature `{}` is not in the list of allowed features: {:?}", k, allowed);
    }
}
```